File: src/lib/scala/Slate.Api/src/main/scala/slate/core/apis/core/Action.scala
/**
<slate_header>
  url: www.slatekit.com
  git: www.github.com/code-helix/slatekit
  org: www.codehelix.co
  author: Kishore Reddy
  copyright: 2016 CodeHelix Solutions Inc.
  license: refer to website and/or github
  about: A Scala utility library, tool-kit and server backend.
  mantra: Simplicity above all else
</slate_header>
  */

package slate.core.apis.core

import slate.common.reflect.ReflectedArg
import slate.core.apis.{Api, ApiAction}

import scala.reflect.runtime.universe.MethodMirror

/**
 * Represents full information about the API Action.
 * This include the metadata and reflection info needed to call the method on the class/ApiBase
 * NOTE: Scala based annotations and reflection are used to get the parameter lists
 *
 * @param name      :  The name of the api action / method name to call
 * @param api       :  The Api annotation put on the method indicating its an api with roles/permissions
 * @param action    :  The ApiAction annotation put on the method indicating its an api action with roles/permissions
 * @param mirror    :  The Scala scala.lang.runtime.universe.MethodMirror to call a method dynamically
 * @param hasArgs   :  Whether the api action has any arguments / parameters ( convenience flag )
 * @param paramList :  A list of the parameters on the method ( name, typename, typeSymbol, position )
 */
class Action(
                      val name      : String            ,
                      val api       : Api               ,
                      val action    : ApiAction         ,
                      val mirror    : MethodMirror      ,
                      val hasArgs   : Boolean           ,
                      val paramList : List[ReflectedArg]
                    )
{
  def isSingleDefaultedArg():Boolean = {
    if(!hasArgs || paramList.size != 1) {
      false
    }
    else {
      val arg = paramList(0)
      arg.isParamDefaulted
    }
  }


  def isSingleArg():Boolean = {
    hasArgs && paramList.size == 1
  }
}

File: src/lib/kotlin/slatekit-apis/src/main/kotlin/slatekit/apis/core/Action.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.apis.core

import slatekit.common.Meta
import slatekit.common.Request
import slatekit.common.ext.tail
import kotlin.reflect.KCallable
import kotlin.reflect.KParameter
import kotlin.reflect.full.createType

/**
 * @param api : Reference to the API associated w/ the action
 * @param member : The callable method associated w/ the action
 * @param name : Name of action which may have a different name than method due to conventions
 * @param desc : Description of the action
 * @param roles : Roles allowed to call this action
 * @param verb : Get/Post verb for Http enabled protocol
 * @param protocol : Protocol associated with the action.
 */
data class Action(
    val member: KCallable<*>,
    val name: String = "",
    val desc: String = "",
    val roles: String = "",
    val verb: String = "*",
    val protocol: String = "*",
    val tag: String = ""
) {
    /**
     * All the parameters of the function, this includes:
     *
     * 1. 0th instance parameter for kotlin
     * 2. a possible Request
     * 3. a possible Meta
     * 4. actual parameters for the method
     */
    private val paramsAll = member.parameters

    /**
     * All the function specific parameters WITHOUT references to
     * the following: These are the parameters that can be
     * discovered, documented, validated against
     *
     * 1. 0th instance parameter for kotlin
     * 2. a possible request/meta paramter
     */
    val paramsUser = filter(member.parameters)

    /**
     * All the parameters that can me mapped over for
     * populating during calls on the CLI / Web
     */
    val params =
            if (paramsAll.size <= 1) listOf()
            else paramsAll.tail()

    /**
     * Whether the action has any arguments.
     */
    val hasArgs = !params.isEmpty()

    fun isSingleDefaultedArg(): Boolean {
        return if (!hasArgs || params.size > 1) {
            false
        } else {
            paramsUser.isNotEmpty() && paramsUser[0].isOptional
        }
    }

    fun isSingleArg(): Boolean = hasArgs && params.size == 1

    companion object {

        @JvmStatic
        val TypeRequest = Request::class.createType()

        @JvmStatic
        val TypeMeta = Meta::class.createType()

        @JvmStatic
        fun filter(args: List<KParameter>): List<KParameter> {
            if (args.isEmpty()) return args
            val finalArgs = args.tail().filter { arg ->
                val type = arg.type
                type != TypeRequest && arg.type != TypeMeta
            }.toList()
            return finalArgs
        }
    }
}
